File: tests/.hooks/typescriptv2/no-zod/src/hooks/customsecurityhook.ts
import { BeforeRequestContext, BeforeRequestHook } from "./types.js";

import { SchemeCustomSchemeAppID } from "../sdk/models/shared/schemecustomschemeappid.js";

export class CustomSecurityHook implements BeforeRequestHook {
  beforeRequest(hookCtx: BeforeRequestContext, request: Request): Request {
    switch (hookCtx.operationID) {
      case "customSchemeAppId": {
        let sec = hookCtx.securitySource;
        if (typeof sec === "function") {
          sec = sec();
        }
        if (!sec) {
          throw new Error("security source is not defined");
        }

        const customSec = sec.customSchemeAppId as SchemeCustomSchemeAppID;
        request.headers.set("X-Security-App-Id", customSec.appId);
        request.headers.set("X-Security-Secret", customSec.secret);

        break;
      }
    }

    return request;
  }
}

File: tests/.hooks/pythonv2/primary/src/openapi/_hooks/customsecurityhook.py
from typing import Union, cast
import httpx

from openapi.models.shared.security import Security

from .types import (
    BeforeRequestContext,
    BeforeRequestHook,
)


class CustomSecurityHook(BeforeRequestHook):
    def before_request(
        self, hook_ctx: BeforeRequestContext, request: httpx.Request
    ) -> Union[httpx.Request, Exception]:
        if hook_ctx.operation_id == "customSchemeAppId":
            if hook_ctx.security_source is None:
                return Exception("security source is None")

            sec = hook_ctx.security_source
            if callable(sec):
                sec = sec()

            security = cast(Security, sec)
            custom_sec = security.custom_scheme_app_id
            if custom_sec is None:
                return Exception("custom_scheme_app_id is not defined")

            request.headers["X-Security-App-Id"] = custom_sec.app_id
            request.headers["X-Security-Secret"] = custom_sec.secret

        return request
